pkg/serviceaccounts/util/helpers.go

```diff
@@ -14,26 +14,56 @@ const (
 	maxSecretPrefixNameLength = maxNameLength - randomLength
 )
 
-func GetDockercfgSecretNamePrefix(serviceAccount *kapi.ServiceAccount) string {
-	return apihelpers.GetName(serviceAccount.Name, "dockercfg-", maxSecretPrefixNameLength)
+func GetDockercfgSecretNamePrefix(serviceAccountName string) string {
+	return apihelpers.GetName(serviceAccountName, "dockercfg-", maxSecretPrefixNameLength)
 }
 
 // GetTokenSecretNamePrefix creates the prefix used for the generated SA token secret. This is compatible with kube up until
 // long names, at which point we hash the SA name and leave the "token-" intact.  Upstream clips the value and generates a random
 // string.
 // TODO fix the upstream implementation to be more like this.
-func GetTokenSecretNamePrefix(serviceAccount *kapi.ServiceAccount) string {
-	return apihelpers.GetName(serviceAccount.Name, "token-", maxSecretPrefixNameLength)
+func GetTokenSecretNamePrefix(serviceAccountName string) string {
+	return apihelpers.GetName(serviceAccountName, "token-", maxSecretPrefixNameLength)
 }
 
-func GetDockercfgSecretNamePrefixV1(serviceAccount *v1.ServiceAccount) string {
-	return apihelpers.GetName(serviceAccount.Name, "dockercfg-", maxSecretPrefixNameLength)
+// IsValidServiceAccountToken returns true if the given secret contains a service account token valid for the given service account
+// TODO, these should collapse onto the upstream method
+func IsValidServiceAccountToken(serviceAccount *kapi.ServiceAccount, secret *kapi.Secret) bool {
```

Comment:
This just pushed your problem around.  It didn't fix it.  Leave this one where it lies and make your external one.